server/mqtt.go

```diff
@@ -4271,32 +4375,39 @@ func mqttDeliverMsgCbQoS0(sub *subscription, pc *client, _ *Account, subject, re
 		return
 	}
 
-	hdr, msg := pc.msgParts(rmsg)
-
 	// This is the client associated with the subscription.
 	cc := sub.client
 
 	// This is immutable
 	sess := cc.mqtt.sess
 
-	// Check the subscription's QoS. This needs to be protected because
-	// the client may change an existing subscription at any time.
-	sess.mu.Lock()
+	// Lock here, otherwise we may be called with sub.mqtt == nil. Ignore
+	// wildcard subscriptions if this subject starts with '$', per Spec
+	// [MQTT-4.7.2-1].
+	sess.subsMu.RLock()
 	subQoS := sub.mqtt.qos
-	isReservedSub := mqttIsReservedSub(sub, subject)
-	sess.mu.Unlock()
+	ignore := mqttMustIgnoreForReservedSub(sub, subject)
+	prmReady := sub.mqtt.prmReady
+	sess.subsMu.RUnlock()
 
-	// We have a wildcard subscription and this subject starts with '$' so ignore per Spec [MQTT-4.7.2-1].
-	if isReservedSub {
+	if ignore {
 		return
 	}
 
+	// Do not proceed until  all retained messages for the sub have been
+	// serialized.
+	if prmReady != nil {
```

Comment:
Re: c.addToPCD(). This is because this map is meant to be used only in the client's readLoop. So there is no locking used there. If we divert messages from incoming readLoop (say pc.kind == ROUTER, etc..), then we can no longer use addToPCD() and will have to find a way to know which "destination" clients we need to flushOutbound/signal_the_writeLoop.
I am not sure at the moment how we will do that (efficiently). One simple approach would be to flush or signal each `cc` in mqttEnqueuePublishMsgTo().

Reply:
(keeping this thread for the discussion of handling all special clients asynchronously)

Reply:
((If we do the special handling, it'd be a separate PR, resolving this thread)